### PR TITLE
Initial attempt at isolating resources between dependencies

### PR DESCRIPTION
**Pull request by stevendanna**
Previously, all resources were loaded into a single resource registry.
Now, each profile context has a resource registry, when a profile's
library is loaded into the profile context, we update the
profile-context-specific resource registry.  This local registry is
then used to populate the execution context that the rules are
evaluated in.

Signed-off-by: Steven Danna steve@chef.io

### CODE REVIEW

File: test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    describe gordon_config do
+      its('version') { should eq('1.0') }
+      its('file_size') { should > 1 }
+    end
+  "
+
+  # Load the configuration file on initialization
+  def initialize
+    @params = {}
+    @path = '/tmp/gordon/config.yaml'
```
Comment (chris-rock) [inline test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb]: we may simplify the resource to just return a version number. Otherwise we are relying on external configuration. eg.

```
def version
   1
end
```

File: test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
+require 'yaml'
+puts "Loading gordon_config from c"
+
+# Custom resource based on the InSpec resource DSL
+class GordonConfig < Inspec.resource(1)
+  name 'gordon_config'
+
+  desc "
```
Comment (chris-rock) [inline test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb]: we do not need desc and example in this test

File: test/unit/mock/profiles/dependencies/profile_a/controls/example.rb [hunk 3]
```diff
@@ -20,3 +20,9 @@
     it { should be_directory }
   end
 end
+
+control 'profilea-2' do
+  describe gordon_config do
+    its('version') { should eq('1.0') }
```
Comment (chris-rock) [inline test/unit/mock/profiles/dependencies/profile_a/controls/example.rb:26]: lets test for `its('version') { should eq 1 }`

File: test/unit/mock/profiles/dependencies/profile_b/controls/example.rb [hunk 4]
```diff
@@ -14,3 +14,9 @@
     it { should be_directory }
   end
 end
+
+control 'profileb-2' do
+  describe gordon_config do
+    its('version') { should eq('1.0') }
```
Comment (chris-rock) [inline test/unit/mock/profiles/dependencies/profile_b/controls/example.rb]: lets test for `its('version') { should eq 2 }`

File: test/unit/mock/profiles/dependencies/inheritance/inspec.yml [hunk 5]
```diff
@@ -10,5 +10,3 @@ depends:
     path: ../profile_a
   - name: profile_b
     path: ../profile_b
-  - name: ssh-hardening
```
Comment (chris-rock) [inline test/unit/mock/profiles/dependencies/inheritance/inspec.yml:13]: any reason why we are removing this?

File: lib/inspec/profile_context.rb [hunk 6]
```diff
@@ -4,17 +4,20 @@
...
   class ProfileContext # rubocop:disable Metrics/ClassLength
     def self.for_profile(profile, backend)
-      new(profile.name, backend, { 'profile' => profile })
+      c = new(profile.name, backend, { 'profile' => profile })
+      c.load_profile_content(profile)
     end
 
-    attr_reader :attributes, :rules, :profile_id
+    attr_reader :attributes, :rules, :profile_id, :resource_registry
+    attr_reader :profile_execution_context
```
Comment (chris-rock) [inline lib/inspec/profile_context.rb]: may it make sense to just use `execution_context`? since we are in the profile context already

File: lib/inspec/resource.rb [hunk 7]
```diff
@@ -3,17 +3,36 @@
 # license: All rights reserved
 # author: Dominik Richter
 # author: Christoph Hartmann
-
+require 'thread'
 require 'inspec/plugins'
 
 module Inspec
   class Resource
+    REGISTRY_LOCK = Mutex.new
```
Comment (chris-rock) [inline lib/inspec/resource.rb]: Why do we need threading and a mutex?

File: lib/inspec/profile_context.rb (hunk 6 above)
Comment (stevendanna) [reply lib/inspec/profile_context.rb]: Based on our discussion I think we'll end up with two context, perhaps:
- `control_eval_context`
- `resource_eval_context`

Comment (chris-rock) [reply lib/inspec/profile_context.rb]: This makes sense to me

File: lib/inspec/resource.rb (hunk 7 above)
Comment (stevendanna) [reply lib/inspec/resource.rb]: Summary of discussion:

We can probably replace this by introducing a profile-specific class that all library files are evaluated against.

File: test/unit/mock/profiles/dependencies/profile_a/controls/example.rb (hunk 3 above)
Comment (stevendanna) [reply test/unit/mock/profiles/dependencies/profile_a/controls/example.rb:26]: Simplified these as suggested.

Comment (chris-rock) [reply test/unit/mock/profiles/dependencies/profile_a/controls/example.rb:26]: This looks great now

File: test/unit/dsl/control_test.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@ def load(content)
       'inspec.yml' => "name: mock",
       'controls/mock.rb' => "control '1' do\n#{content}\nend\n",
     }
-    opts = { test_collector: Inspec::RunnerMock.new }
+    opts = { test_collector: Inspec::RunnerMock.new, backend: Inspec::Backend.create({}) }
```
Comment (chris-rock) [inline test/unit/dsl/control_test.rb]: wouldn't that require `Inspec::Backend.create({ backend: 'mock' })`?

File: lib/inspec/runner.rb [hunk 9]
```diff
@@ -102,61 +120,15 @@ def load_attributes(options)
...
-    # considered by this call path are:
-    #
-    # - `:ignore_supports`: A boolean that controls whether to ignore
-    #   the profile's metadata regarding supported Inspec versions or
-    #   platforms.
-    #
-    # - `:controls`: An array of controls to run from this
-    #    profile. Any returned rules that are not part of these
-    #    controls will be filtered before being passed to @test_collector.
+    # @params _opts [Hash] Unused, but still here to avoid breaking kitchen-inspec
```
Comment (chris-rock) [inline lib/inspec/runner.rb:124]: We should add a ticket in kitchen-inspec as a follow-up to fix that

File: test/unit/mock/profiles/dependencies/inheritance/inspec.yml (hunk 5 above)
Comment (stevendanna) [reply test/unit/mock/profiles/dependencies/inheritance/inspec.yml:13]: @chris-rock Mostly because the fetching wasn't the main thing we were working so the ssh output was really just cluttering up what we wanted to see.

**Comment by vjeffrey** [comment]
:+1:

File: lib/inspec/base_cli.rb [hunk 10]
```diff
@@ -70,7 +70,7 @@ def run_tests(targets, opts)
       o[:logger].level = get_log_level(o.log_level)
 
       runner = Inspec::Runner.new(o)
-      targets.each { |target| runner.add_target(target, opts) }
+      targets.each { |target| runner.add_target(target) }
```
Comment (chris-rock) [inline lib/inspec/base_cli.rb:73]: 👍

File: lib/inspec/control_eval_context.rb [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+          @require_loader = require_loader
+          @skip_profile = false
+        end
+
+        define_method :title do |arg|
+          profile_context_owner.set_header(:title, arg)
+        end
+
+        def to_s
+          "Profile Context Run #{profile_name}"
```
Comment (chris-rock) [inline lib/inspec/control_eval_context.rb]: Is this really representing a full profile context? Its only representing one section aka file in a profile

File: lib/inspec/control_eval_context.rb [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+          profile_id
+        end
+
+        define_method :control do |*args, &block|
+          id = args[0]
+          opts = args[1] || {}
+          register_control(rule_class.new(id, profile_id, opts, &block))
+        end
+
+        define_method :describe do |*args, &block|
```
Comment (chris-rock) [inline lib/inspec/control_eval_context.rb:82]: do you think we should add a comment that this is simulating the rspec describe and implicitly generate a control?

File: lib/inspec/control_eval_context.rb [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+          res = nil
+          rule = rule_class.new(id, profile_id, {}) do
+            res = describe(*args, &block)
+          end
+          register_control(rule, &block)
+
+          res
+        end
+
+        define_method :add_resources do |context|
```
Comment (chris-rock) [inline lib/inspec/control_eval_context.rb:95]: the methods `add_resources`, `add_subcontext` and `register_control` should not be available to our DSL right?

File: lib/inspec/dependencies/dependency_set.rb [hunk 14]
```diff
@@ -18,14 +18,22 @@ class DependencySet
     # @param cwd [String] Current working directory for relative path includes
     # @param vendor_path [String] Path to the vendor directory
     #
-    def self.from_lockfile(lockfile, cwd, vendor_path)
+    def self.from_lockfile(lockfile, cwd, vendor_path, backend)
```
Comment (chris-rock) [inline lib/inspec/dependencies/dependency_set.rb:21]: param is not defined in comments

File: lib/inspec/dependencies/dependency_set.rb [hunk 15]
```diff
@@ -18,14 +18,22 @@ class DependencySet
...
       vendor_index = VendorIndex.new(vendor_path)
       dep_tree = lockfile.deps.map do |dep|
-        Inspec::Requirement.from_lock_entry(dep, cwd, vendor_index)
+        Inspec::Requirement.from_lock_entry(dep, cwd, vendor_index, backend)
```
Comment (chris-rock) [inline lib/inspec/dependencies/dependency_set.rb:24]: Somehow it feels strange that the dependency resolution needs a backend? Why is that required?

File: lib/inspec/dsl_shared.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
+# encoding: utf-8
+module Inspec
+  #
+  # Contains methods we would like in multiple DSL
+  #
+  module DSL
+    module RequireOverride
```
Comment (chris-rock) [inline lib/inspec/dsl_shared.rb:7]: nice separation

File: lib/inspec/control_eval_context.rb (hunk 13 above)
Comment (stevendanna) [reply lib/inspec/control_eval_context.rb:95]: We could hide these from the API directly via a layer of indirection.  If we want them to be completely hidden, I think we would need the Inspec::DSL code to change a bit more substantially.  Happy to do either.

File: lib/inspec/dependencies/dependency_set.rb (hunk 15 above)
Comment (stevendanna) [reply lib/inspec/dependencies/dependency_set.rb:24]: Yeah, I agree this is gross.  The reason we started threading this through everything is that we need the backend anywhere we might instantiate a ProfileContext.  Since we instantiate ProfileContext's as we are building Inpec::Profile's, we now need it all over the place.

Comment (chris-rock) [reply lib/inspec/dependencies/dependency_set.rb:24]: As discussed, lets improve that in our next iterations

**Comment by chris-rock** [comment]
Great improvements @stevendanna and @vjeffrey

**Comment by chris-rock** [comment]
#958
